src/psyclone/transformations.py

```diff
@@ -1138,3 +1178,143 @@ def apply(self, node, location, position="before"):
+      raise an exception if this is not the case.
+
+    * This transformation can not be applied to a loop containing a
+      reduction and will again raise an exception if this is the case.
+
+    * This transformation can only be used to add redundant
+      computation to a loop, not to remove it, however it does allow
+      redundant computation depths to be reduced (if this is ever
+      required).'''
+
+    def __str__(self):
+        return "Change iteration space to perform redundant computation"
+
+    @property
+    def name(self):
+        ''' Returns the name of this transformation as a string '''
+        return "RedundantComputation"
+
+    def _validate(self, node, depth):
+        '''Perform various checks to ensure that it is valid to apply the
+        RedundantComputation transformation to the supplied node'''
+
+        # check node is a loop
+        from psyGen import Loop
+        if not isinstance(node, Loop):
+            raise TransformationError(
+                "In the DynamoRedundantComputation transformation apply "
+                "method the first argument is not a Loop")
+
+        # check loop's parent is the schedule otherwise halo exchange
+        # placement fails. The only current example when this would be
+        # the case is when directives have been added.
+        from psyGen import Schedule
+        if not isinstance(node.parent, Schedule):
+            raise TransformationError(
+                "In the DynamoRedundantComputation transformation apply "
+                "method the parent must be the schedule, but found "
+                "{0}".format(type(node.parent)))
+
+        import config
+        if not config.DISTRIBUTED_MEMORY:
+            raise TransformationError(
+                "In the DynamoRedundantComputation transformation apply "
+                "method distributed memory must be switched on")
+
+        # loop must iterate over cells or dofs. This currently
+        # precludes loops over colours. Note, an empty loop_type
+        # iterates over cells
+        if node.loop_type not in ["", "dofs"]:
+                raise TransformationError(
+                    "In the DynamoRedundantComputation transformation apply "
+                    "method the loop must iterate over cells or dofs, but "
+                    "found '{0}'".format(node.loop_type))
+
+        if depth is None:
+            if node.upper_bound_name in ["cell_halo", "dof_halo"]:
+                if not node.upper_bound_index:
+                    raise TransformationError(
+                        "In the DynamoRedundantComputation transformation "
+                        "apply method the loop is already set to the maximum "
+                        "halo depth so this transformation does nothing")
+                for call in node.calls():
+                    for arg in call.arguments.args:
+                        if arg.stencil:
+                            raise TransformationError(
```

Comment:
I guess this is a question for the users. We could do what you say and have an upper bound that is max_depth-stencil_size, but that would need a runtime error check in case stencil_size was too large. My opinion is that we want to avoid max_depth where possible as it is something we can't reason about when optimising. To be honest I'd ban max_depth if I could!